### 2021_11_12_3

### RNN Basic

In [2]:
import tensorflow as tf
import keras
import numpy as np

In [4]:
for i in range(10) :
    lst = list(range(i, i + 5))
    print(lst)

[0, 1, 2, 3, 4]
[1, 2, 3, 4, 5]
[2, 3, 4, 5, 6]
[3, 4, 5, 6, 7]
[4, 5, 6, 7, 8]
[5, 6, 7, 8, 9]
[6, 7, 8, 9, 10]
[7, 8, 9, 10, 11]
[8, 9, 10, 11, 12]
[9, 10, 11, 12, 13]


In [6]:
X = []
Y = []

for i in range(10) :
    lst = list(range(i, i + 5))
    X.append([[c / 10] for c in lst])
    Y.append((i + 5) / 10)
    
X = np.array(X)
Y = np.array(Y)

print("X shape :", X.shape, " Y shape :",  Y.shape)
print(X[0], Y[0])
print(X[1], Y[1])

X shape : (10, 5, 1)  Y shape : (10,)
[[0. ]
 [0.1]
 [0.2]
 [0.3]
 [0.4]] 0.5
[[0.1]
 [0.2]
 [0.3]
 [0.4]
 [0.5]] 0.6


In [8]:
# 전체 데이터 확인
for i in range(len(X)) :
    print(X[i], Y[i])
print("X shape :", X.shape)
print("Y shape :", Y.shape)

[[0. ]
 [0.1]
 [0.2]
 [0.3]
 [0.4]] 0.5
[[0.1]
 [0.2]
 [0.3]
 [0.4]
 [0.5]] 0.6
[[0.2]
 [0.3]
 [0.4]
 [0.5]
 [0.6]] 0.7
[[0.3]
 [0.4]
 [0.5]
 [0.6]
 [0.7]] 0.8
[[0.4]
 [0.5]
 [0.6]
 [0.7]
 [0.8]] 0.9
[[0.5]
 [0.6]
 [0.7]
 [0.8]
 [0.9]] 1.0
[[0.6]
 [0.7]
 [0.8]
 [0.9]
 [1. ]] 1.1
[[0.7]
 [0.8]
 [0.9]
 [1. ]
 [1.1]] 1.2
[[0.8]
 [0.9]
 [1. ]
 [1.1]
 [1.2]] 1.3
[[0.9]
 [1. ]
 [1.1]
 [1.2]
 [1.3]] 1.4
X shape : (10, 5, 1)
Y shape : (10,)


### 모델 구축
* SimpleRNN
    * return_sequences : 기본값(False)
        * False : 마지막 상태만 출력
        * True : 모든 지점의 은닉 상태 출력
    * return_state : 기본값(False)
        * True : return_sequences의 값과 상관없이 마지막 은닉 상태를 출력

In [15]:
model = tf.keras.Sequential([
    tf.keras.layers.SimpleRNN(units = 10, 
                              return_sequences = False,
                              input_shape = [5, 1]), 
    tf.keras.layers.Dense(1)
])

In [16]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn_2 (SimpleRNN)    (None, 10)                120       
                                                                 
 dense_2 (Dense)             (None, 1)                 11        
                                                                 
Total params: 131
Trainable params: 131
Non-trainable params: 0
_________________________________________________________________


In [17]:
model.compile(optimizer = "adam", loss = "mse")

model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn_2 (SimpleRNN)    (None, 10)                120       
                                                                 
 dense_2 (Dense)             (None, 1)                 11        
                                                                 
Total params: 131
Trainable params: 131
Non-trainable params: 0
_________________________________________________________________


### 파라미터 개수
* 파라미터 아웃 * (파라미터 아웃 + 차원 + 1(바이어스))
* 10 (10 1 + 1) = 120

In [18]:
model.fit(X, Y, epochs = 50, verbose = 1)

Epoch 1/50
1/1 [==============================] - 1s 577ms/step - loss: 2.6768
Epoch 2/50
1/1 [==============================] - 0s 4ms/step - loss: 2.5926
Epoch 3/50
1/1 [==============================] - 0s 3ms/step - loss: 2.5097
Epoch 4/50
1/1 [==============================] - 0s 3ms/step - loss: 2.4282
Epoch 5/50
1/1 [==============================] - 0s 4ms/step - loss: 2.3481
Epoch 6/50
1/1 [==============================] - 0s 4ms/step - loss: 2.2694
Epoch 7/50
1/1 [==============================] - 0s 6ms/step - loss: 2.1922
Epoch 8/50
1/1 [==============================] - 0s 5ms/step - loss: 2.1166
Epoch 9/50
1/1 [==============================] - 0s 5ms/step - loss: 2.0425
Epoch 10/50
1/1 [==============================] - 0s 5ms/step - loss: 1.9699
Epoch 11/50
1/1 [==============================] - 0s 4ms/step - loss: 1.8990
Epoch 12/50
1/1 [==============================] - 0s 4ms/step - loss: 1.8297
Epoch 13/50
1/1 [==============================] - 0s 4ms/step - loss: 

In [19]:
print(Y) # 실제값
pred = model.predict(X)
pred

[0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4]


array([[0.41972467],
       [0.50236946],
       [0.55135244],
       [0.5736221 ],
       [0.576312  ],
       [0.5651973 ],
       [0.5445627 ],
       [0.5175112 ],
       [0.48629412],
       [0.4525499 ]], dtype=float32)